tfsdk/tftypes_value.go

```diff
@@ -0,0 +1,221 @@
+package tfsdk
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+)
+
+// createParentValue ensures that the given parent value can have children
+// values upserted. If the parent value is known and not null, it is returned
+// without modification. A null Object or Tuple is converted to known with null
+// children. An unknown Object or Tuple is converted to known with unknown
+// children. List, Map, and Set are created with empty elements.
+//
+// The parentType parameter ensures that the value will match the current
+// schema, not what is currently stored in the parentValue.
```

Comment:
I think we would actually like this to actively _not_ be the case; the schema represents type constraints (could be DynamicPseudoType, could be an object with optional attributes) but the value represents a concrete type (any non-DynamicPseudoType type, could be an object without optional attributes, respectively). So when modifying a value, I think we want to stick with the value's interpretation of what the type is supposed to be, not the schema's, so the value can stay internally consistent and match its type. Technically, it may even panic in situations where the value and type of the value start deviating.